packages/beacon-node/src/chain/produceBlock/produceBlockBody.ts

```diff
@@ -303,110 +276,151 @@ export async function produceBlockBody<T extends BlockType>(
-            blobsResult = {type: BlobsResultType.produced, contents, blockHash};
+        if (prepType !== PayloadPreparationType.Cached) {
+          // Wait for 500ms to allow EL to add some txs to the payload
+          // the pitfalls of this have been put forward here, but 500ms delay for block proposal
+          // seems marginal even with unhealthy network
+          //
+          // See: https://discord.com/channels/595666850260713488/892088344438255616/1009882079632314469
+          await sleep(PAYLOAD_GENERATION_TIME_MS);
+        }
 
-            Object.assign(logMeta, {blobs: blobsBundle.commitments.length});
-          } else {
-            blobsResult = {type: BlobsResultType.preDeneb};
-          }
+        const payloadRes = await this.executionEngine.getPayload(fork, payloadId);
 
-          if (ForkSeq[fork] >= ForkSeq.electra) {
-            if (executionRequests === undefined) {
-              throw Error(`Missing executionRequests response from getPayload at fork=${fork}`);
-            }
-            (blockBody as electra.BeaconBlockBody).executionRequests = executionRequests;
-          }
-        }
-      } catch (e) {
+        return {...prepareRes, ...payloadRes};
+      })().catch((e) => {
+        // catch payload fetch here, because there is still a recovery path possible if we
+        // are pre-merge. We don't care the same for builder segment as the execution block
+        // will takeover if the builder flow was activated and errors
         this.metrics?.blockPayload.payloadFetchErrors.inc();
-        // ok we don't have an execution payload here, so we can assign an empty one
-        // if pre-merge
 
         if (!isMergeTransitionComplete(currentState as CachedBeaconStateBellatrix)) {
           this.logger?.warn(
             "Fetch payload from the execution failed, however since we are still pre-merge proceeding with an empty one.",
             {},
             e as Error
           );
-          (blockBody as BeaconBlockBody<ForkPostBellatrix>).executionPayload =
-            sszTypesFor(fork).ExecutionPayload.defaultValue();
-          blobsResult = {type: BlobsResultType.preDeneb};
-          executionPayloadValue = BigInt(0);
+          // ok we don't have an execution payload here, so we can assign an empty one
+          // if pre-merge
+          return {
+            isPremerge: true as const,
+            executionPayload: sszTypesFor(fork).ExecutionPayload.defaultValue(),
+            executionPayloadValue: BigInt(0),
+          };
+        }
+        // since merge transition is complete, we need a valid payload even if with an
+        // empty (transactions) one. defaultValue isn't gonna cut it!
+        throw e;
+      });
+
+      const [engineRes, commonBlockBody] = await Promise.all([
+        enginePromise,
+        commonBlockBodyFn?.() ?? produceCommonBlockBody.call(this, blockType, currentState, blockAttr),
```

Comment:
what if `commonBlockBodyFn()` is called even before `enginePromise` by the builder flow?

Reply:
ah right this is a problem, we would need to trigger it once both requests are sent which I don't see with this solution how we can achieve this

I think passing the function down the call stack is not ideal, we need to decouple assembling the block from sending builder / execution request